### PR TITLE
feat(connector): [SQUARE] Implement webhooks

### PR DESCRIPTION
**Pull request by deepanshu-iiitu**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Square is a payment connector from North America.
Add webhooks support for Square.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Add Webhooks for Payments and Refunds flow for new connector square.

## How did you test it?

Testing was completed using webhooks.site

## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/square.rs [hunk 1]
```diff
@@ -787,24 +790,101 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+                }
+                square::SquarePaymentStatus::Failed => {
+                    api::IncomingWebhookEvent::PaymentIntentFailure
+                }
+                square::SquarePaymentStatus::Approved
+                | square::SquarePaymentStatus::Canceled
+                | square::SquarePaymentStatus::Pending
+                | square::SquarePaymentStatus::Processing => {
+                    api::IncomingWebhookEvent::EventNotSupported
+                }
```
Comment (ArjunKarthik) [inline crates/router/src/connector/square.rs:869]: Let's move this to from implementation

File: crates/router/src/connector/square.rs [hunk 2]
```diff
@@ -787,24 +790,101 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+            },
+            square::SquareWebhookObject::Refund(refund_data) => match refund_data.status {
+                square::RefundStatus::Completed => api::IncomingWebhookEvent::RefundSuccess,
+                square::RefundStatus::Failed => api::IncomingWebhookEvent::RefundFailure,
+                square::RefundStatus::Pending | square::RefundStatus::Processing => {
+                    api::IncomingWebhookEvent::EventNotSupported
```
Comment (ArjunKarthik) [inline crates/router/src/connector/square.rs:875]: same here

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/square.rs (hunk 1 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square.rs:872]: Done

File: crates/router/src/connector/square.rs (hunk 2 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square.rs:878]: Done

File: crates/router/src/connector/square.rs [hunk 3]
```diff
@@ -787,24 +790,101 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+        })
     }
 
     fn get_webhook_resource_object(
         &self,
-        _request: &api::IncomingWebhookRequestDetails<'_>,
+        request: &api::IncomingWebhookRequestDetails<'_>,
     ) -> CustomResult<serde_json::Value, errors::ConnectorError> {
-        Err(errors::ConnectorError::WebhooksNotImplemented).into_report()
+        let reference_object: serde_json::Value = serde_json::from_slice(request.body)
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square.rs]: Don

File: crates/router/src/connector/square.rs [hunk 4]
```diff
@@ -809,24 +811,93 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+        let webhook_body: square::SquareWebhookBody = request
+            .body
+            .parse_struct("SquareWebhookBody")
+            .change_context(errors::ConnectorError::WebhookReferenceIdNotFound)?;
+
+        match webhook_body.data.object {
+            square::SquareWebhookObject::Payment(_) => {
+                Ok(api_models::webhooks::ObjectReferenceId::PaymentId(
+                    api_models::payments::PaymentIdType::ConnectorTransactionId(
+                        webhook_body.data.id,
```
Comment (ArjunKarthik) [inline crates/router/src/connector/square.rs:857]: do we get back our payment_attempt_id in web hooks? if so let's use that

File: crates/router/src/connector/square.rs [hunk 5]
```diff
@@ -809,24 +811,93 @@ impl ConnectorIntegration<api::RSync, types::RefundsData, types::RefundsResponse
...
+                    ),
+                ))
+            }
+            square::SquareWebhookObject::Refund(_) => {
+                Ok(api_models::webhooks::ObjectReferenceId::RefundId(
+                    api_models::webhooks::RefundIdType::ConnectorRefundId(webhook_body.data.id),
```
Comment (ArjunKarthik) [inline crates/router/src/connector/square.rs:863]: do we get back our refund_id in web hooks? if so let's use that

**Review by ArjunKarthik — COMMENTED** [review]
Does connector support disputes?

File: crates/router/src/connector/square.rs (hunk 4 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square.rs:857]: No. We get the payment id assigned by connector square as part of Webhooks. Our ID is just used as the Idempotency key.

File: crates/router/src/connector/square.rs (hunk 5 above)
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square.rs:863]: No. We get the refund id assigned by connector square as part of Webhooks. Our ID is just used as the Idempotency key.

**Review by ArjunKarthik — APPROVED** [review]

**Comment by sai-harsha-vardhan** [comment]
> How did you test it?
> Testing was completed using webhooks.site

Can you please elaborate the testing process here? How did you ensure following things?
1. Source verification is happening as expected
2. Payment status updation when source verification is successful

File: crates/router/src/connector/square/transformers.rs [hunk 6]
```diff
@@ -445,3 +445,42 @@ pub struct SquareErrorDetails {
...
+
+impl From<SquareWebhookObject> for api::IncomingWebhookEvent {
+    fn from(item: SquareWebhookObject) -> Self {
+        match item {
+            SquareWebhookObject::Payment(payment_data) => match payment_data.status {
+                SquarePaymentStatus::Completed => Self::PaymentIntentSuccess,
+                SquarePaymentStatus::Failed => Self::PaymentIntentFailure,
+                SquarePaymentStatus::Approved
+                | SquarePaymentStatus::Canceled
+                | SquarePaymentStatus::Pending => Self::EventNotSupported,
```
[the remark this answers is not in the document]
Comment (deepanshu-iiitu) [reply crates/router/src/connector/square/transformers.rs]: Done

**Review by sai-harsha-vardhan — APPROVED** [review]

**Review by ArjunKarthik — APPROVED** [review]

**Review by SanchithHegde — APPROVED** [review]

**Pull request merged**